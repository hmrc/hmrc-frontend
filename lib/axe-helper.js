const { configureAxe } = require('jest-axe');

const axe = configureAxe({
  rules: {
    // As we're testing incomplete HTML fragments, we don't expect there to be a
    // skip link.
    'skip-link': { enabled: false },
  },
});

module.exports = axe;
