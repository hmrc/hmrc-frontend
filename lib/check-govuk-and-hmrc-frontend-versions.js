const fs = require('fs');
const path = require('path');

const projectDir = process.env.INIT_CWD;
const hmrcFrontendDir = path.join(projectDir, 'node_modules/hmrc-frontend/');
let govukFrontendVersion = '';

try {
  govukFrontendVersion = JSON.parse(fs.readFileSync(path.join(hmrcFrontendDir, 'package.json'), 'utf8')).dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
} catch (err) {
  process.exit(0); // Exit if no package.json is found under project_path/node_modules/hmrc-frontend
}

const govukPrototypeKitPackagePath = path.join(projectDir, 'package.json');
const govukPrototypeKitConfig = JSON.parse(fs.readFileSync(govukPrototypeKitPackagePath, 'utf8'));
const currentGovukFrontendVersion = govukPrototypeKitConfig.dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
if (currentGovukFrontendVersion >= govukFrontendVersion) {
  process.exit(0);
} else {
  govukPrototypeKitConfig.dependencies['govuk-frontend'] = govukFrontendVersion;
  // eslint-disable-next-line no-console
  console.warn(`
  Hello friend, looks like you're on
  govuk-frontend v${currentGovukFrontendVersion}, 
  but hmrc-frontend requires at least v${govukFrontendVersion}.
  This error is just a heads up that hmrc-frontend might not
  work as intended on your current version.
    - PlatUI
  -------------------------------------------
          \\   ^__^ 
           \\  (oo)_______
              (__)       )\\/\\
                  ||----w |
                  ||     ||
  - PlatMooI the PlatUI Helper Cow
  `);

  process.exit(1);
}
