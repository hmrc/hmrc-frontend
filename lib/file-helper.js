const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const Remarkable = require('remarkable');

const configPaths = require('../config/paths.json');

const childDirectories = (dir) => fs.readdirSync(dir)
  .filter((file) => {
    const filePath = path.join(configPaths.components, file);
    const isDirectory = fs.statSync(filePath).isDirectory();
    const hasComponentConfig = fs.existsSync(`${filePath}/${file}.yaml`);
    return isDirectory && hasComponentConfig;
  });

// Generate component list from source directory, excluding anything that's not
// a directory (for example, .DS_Store files)
exports.allComponents = childDirectories(configPaths.components);

// Read the contents of a file from a given path
const readFileContents = (filePath) => fs.readFileSync(filePath, 'utf8');

exports.readFileContents = readFileContents;

const getComponentData = (componentName) => {
  const yamlPath = path.join(configPaths.components, componentName, `${componentName}.yaml`);
  return yaml.safeLoad(fs.readFileSync(yamlPath, 'utf8'), { json: true });
};

exports.getComponentData = getComponentData;

const md = new Remarkable();

const govukify = (html) => html
  .replace(/<p>/g, '<p class="govuk-body">')
  .replace(/<h1/g, '<h1 class="govuk-heading-l"')
  .replace(/<h2/g, '<h2 class="govuk-heading-m"')
  .replace(/<h3/g, '<h3 class="govuk-heading-s"')
  .replace(/<ul/g, '<ul class="govuk-list govuk-list--bullet govuk-list--spaced"')
  .replace(/<a /g, '<a class="govuk-link" ');

const getComponentReadme = (componentPath) => {
  const readme = path.join(configPaths.components, componentPath, 'README.md');
  const markdown = readFileContents(readme);
  const html = md.render(markdown);
  return govukify(html);
};

exports.getComponentReadme = getComponentReadme;
