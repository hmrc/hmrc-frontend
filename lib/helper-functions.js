// Convert component name to macro name
//
// This helper function takes a component name and returns the corresponding
// macro name.
//
// Component names are lowercase, dash-separated strings (button, date-input),
// whilst macro names have a `govuk` prefix and are camel cased (govukButton,
// govukDateInput).
const componentNameToMacroName = (componentName) => {
  const macroName = componentName
    .toLowerCase()
    .split('-')
    // capitalize each 'word'
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join('');

  return `hmrc${macroName}`;
};
exports.componentNameToMacroName = componentNameToMacroName;

exports.componentNameToComponentDirectory = (componentName) => [componentName].join('-'
  + '');
