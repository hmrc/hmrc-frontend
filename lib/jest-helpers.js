const path = require('path');
const fs = require('fs');
const nunjucks = require('nunjucks');
const cheerio = require('cheerio');
const yaml = require('js-yaml');
require('jest-axe/extend-expect');

const configPaths = require('../config/paths.json');
const { componentNameToMacroName } = require('./helper-functions');

nunjucks.configure(
  [
    configPaths.components,
    path.join(configPaths.govukFrontend, 'dist'),
    path.join(configPaths.govukFrontend, 'dist/govuk/components'),
  ],
  { trimBlocks: true, lstripBlocks: true },
);

/**
 * Render a component's macro for testing
 * @param {string} componentName
 * @param {string} params parameters that are used in the component macro
 * @param {any} children any child components or text, pass the children to the macro
 * @returns {function} returns cheerio (jQuery) instance of the macro for easy DOM querying
 */
function render(componentName, params, children = false) {
  if (typeof params === 'undefined') {
    throw new Error('Parameters passed to `render` should be an object but are undefined');
  }

  const macroName = componentNameToMacroName(componentName);
  const macroParams = JSON.stringify(params, null, 2);

  let macroString = `{%- from "${componentName}/macro.njk" import ${macroName} -%}`;

  // If we're nesting child components or text, pass the children to the macro
  // using the 'caller' Nunjucks feature
  if (children) {
    macroString += `{%- call ${macroName}(${macroParams}) -%}${children}{%- endcall -%}`;
  } else {
    macroString += `{{- ${macroName}(${macroParams}) -}}`;
  }

  const output = nunjucks.renderString(macroString);

  return cheerio.load(output);
}

/**
 * Get examples from a component's metadata file
 * @param {string} componentName
 * @returns {object} returns object that includes all examples at once
 */
function getExamples(componentName) {
  const file = fs.readFileSync(
    path.join(configPaths.components, componentName, `${componentName}.yaml`),
    'utf8',
  );

  const docs = yaml.safeLoad(file);

  const examples = {};

  docs.examples.forEach((example) => {
    examples[example.name] = example.data;
  });

  return examples;
}

/**
 * Get the raw HTML representation of a component, and remove any other
 * child elements that do not match the component.
 * Relies on B.E.M naming ensuring that child components relating to a component
 * are namespaced.
 * @param {function} $ requires an instance of cheerio (jQuery) that includes the
 * rendered component.
 * @param {string} className the top level class 'Block' in B.E.M terminology
 * @returns {string} returns HTML
 */
function htmlWithClassName($, className) {
  const $component = $(className);
  const classSelector = className.replace('.', '');
  // Remove all other elements that do not match this component
  $component.find(`[class]:not([class^=${classSelector}])`).remove();
  return $.html($component);
}

module.exports = { render, getExamples, htmlWithClassName };
