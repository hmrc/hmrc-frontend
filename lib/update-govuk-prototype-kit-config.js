const path = require('path');
const fs = require('fs');

const projectDir = path.dirname(__dirname);
const govukFrontendVersion = JSON.parse(fs.readFileSync(path.join(projectDir, 'package.json'), 'utf8')).dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
const govukPrototypeKitConfigPath = path.join(projectDir, 'src', 'govuk-prototype-kit.config.json');
const govukPrototypeKitConfig = JSON.parse(fs.readFileSync(govukPrototypeKitConfigPath, 'utf8'));
govukPrototypeKitConfig.pluginDependencies = [
  {
    packageName: 'govuk-frontend',
    minVersion: govukFrontendVersion,
  },
];
fs.writeFileSync(govukPrototypeKitConfigPath, JSON.stringify(govukPrototypeKitConfig, null, 2));
