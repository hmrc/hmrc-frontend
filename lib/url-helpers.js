import { ports } from '../config/paths.json';

export function examplePreview(example) {
  return `http://localhost:${ports.app}/components/${example.replace(/^\/+|\/+$/g, '')}/preview`;
}
