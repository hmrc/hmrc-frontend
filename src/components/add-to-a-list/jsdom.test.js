/* eslint-env jest */

import { render, getExamples } from '../../../lib/jest-helpers';

const examples = getExamples('add-to-a-list');

const getTableItems = ($) => {
  const $rows = $('.hmrc-list-with-actions');
  const $changeLinks = $('.govuk-summary-list__actions-list-item:first-of-type', $rows);
  const $removeLinks = $('.govuk-summary-list__actions-list-item:last-of-type', $rows);

  const listItems = {};
  listItems.identifiers = [];
  $('.govuk-summary-list__key', $rows).each((index, element) => {
    listItems.identifiers.push($(element).text().trim());
  });
  listItems.changeLinkText = $changeLinks.find('[aria-hidden="true"]').eq(0).text().trim();
  listItems.removeLinkText = $removeLinks.find('[aria-hidden="true"]').eq(0).text().trim();
  listItems.ariaChangeText = $changeLinks.find('.govuk-visually-hidden').eq(0).text().trim();
  listItems.ariaRemoveText = $removeLinks.find('.govuk-visually-hidden').eq(0).text().trim();

  return listItems;
};

describe('Add to a list', () => {
  it('should be rendered using a structure accessible for JAWS users', () => {
    // https://github.com/hmrc/design-patterns/issues/31#issuecomment-799628620

    const $ = render('add-to-a-list', examples['multiple-specific-items']);

    expect($('.hmrc-list-with-actions').first().html()).toMatchSnapshot();
  });

  describe('by default', () => {
    const $ = render('add-to-a-list', examples.default);
    const $heading = $('h1');
    const $legend = $('fieldset legend');

    it('Has the correct no-items form of heading', () => {
      expect($heading.text().trim()).toBe('You have not added any items');
    });

    it('has the default legend text', () => {
      expect($legend.text().trim()).toBe('Do you need to add another item?');
    });
  });

  describe('with an empty list in welsh', () => {
    const $ = render('add-to-a-list', examples['empty-list-welsh']);
    const $heading = $('h1');
    const $legend = $('fieldset legend');

    it('Has the correct no-items form of heading in welsh', () => {
      expect($heading.text().trim()).toBe('Nid ydych wedi ychwanegu unrhyw eitemau');
    });
    it('has the default legend text in welsh', () => {
      expect($legend.text().trim()).toBe('Oes angen i chi ychwanegu eitem arall?');
    });
  });

  describe('with one item', () => {
    const $ = render('add-to-a-list', examples['single-generic-item']);
    const $heading = $('h1');
    const $rows = $('.govuk-summary-list__row');

    it('has the correct text for 1 item', () => {
      expect($heading.text()).toBe('You have added 1 item');
    });

    it('Contains a list of 1 item', () => {
      const listItems = getTableItems($);

      expect($rows.length).toBe(1);
      expect(listItems.identifiers[0]).toBe('item one');
      expect(listItems.changeLinkText).toBe('Change');
      expect(listItems.removeLinkText).toBe('Remove');
      expect(listItems.ariaChangeText).toBe('Change item one');
      expect(listItems.ariaRemoveText).toBe('Remove item one from the list');
    });
  });

  describe('with one item in welsh', () => {
    const $ = render('add-to-a-list', examples['single-generic-welsh-item']);
    const $heading = $('h1');
    const $rows = $('.hmrc-list-with-actions');

    it('Has the correct singular welsh heading', () => {
      expect($heading.text()).toBe('Rydych wedi ychwanegu 1 eitem');
    });

    it('Contains a list of 1 item', () => {
      const listItems = getTableItems($);

      expect($rows.length).toBe(1);
      expect(listItems.identifiers[0]).toBe('item one');
      expect(listItems.changeLinkText).toBe('Newid');
      expect(listItems.removeLinkText).toBe('Dileu');
      expect(listItems.ariaChangeText).toBe('Newid item one');
      expect(listItems.ariaRemoveText).toBe('Dileu’r item one o’r rhestr');
    });
  });

  describe('with two named items', () => {
    const $ = render('add-to-a-list', examples['multiple-specific-items']);
    const $heading = $('h1');
    const $rows = $('.hmrc-list-with-actions');
    const $legend = $('fieldset legend');
    const $form = $('form');
    it('Uses "directors" as the plural item name', () => {
      expect($heading.text()).toContain('directors');
    });

    it('Uses "2" as the count value', () => {
      expect($heading.text()).toContain('2 directors');
    });

    it('Contains a list of 2 directors', () => {
      const $identifiers = $('.govuk-summary-list__key', $rows);
      const $changeLinks = $('.govuk-summary-list__actions-list-item:first-of-type .govuk-visually-hidden', $rows);
      const $removeLinks = $('.govuk-summary-list__actions-list-item:last-of-type .govuk-visually-hidden', $rows);

      expect($rows.children().length).toBe(2);
      expect($identifiers.eq(0).text()).toContain('Director One');
      expect($changeLinks.eq(0).text()).toContain('Change Director One');
      // TODO: add test for the change url
      expect($removeLinks.eq(0).text()).toContain('Remove Director One from the list');
      // TODO: add test for the remove url

      expect($identifiers.eq(1).text()).toContain('Director Two');
      expect($changeLinks.eq(1).text()).toContain('Change Director Two');
      expect($removeLinks.eq(1).text()).toContain('Remove Director Two from the list');
    });

    it('has the item type in the legend text', () => {
      expect($legend.text()).toContain('Do you need to add another director?');
    });

    it('has an action set on the form', () => {
      expect($form.attr('action')).toBe('#add-director');
    });
  });

  describe('with two unnamed items in Welsh', () => {
    const $ = render('add-to-a-list', examples['multiple-generic-welsh-items']);
    const $heading = $('h1');
    const $rows = $('.hmrc-list-with-actions');
    const $legend = $('fieldset legend');
    const $form = $('form');
    it('Uses "o eitemau" as the plural item name', () => {
      expect($heading.text()).toBe('Rydych wedi ychwanegu 2 o eitemau');
    });

    it('Uses "2" as the count value', () => {
      expect($heading.text()).toContain('2 o eitemau');
    });

    it('Contains a list of 2 items', () => {
      const $identifiers = $('.govuk-summary-list__key', $rows);
      const $changeLinks = $('.govuk-summary-list__actions-list-item:first-of-type .govuk-visually-hidden', $rows);
      const $removeLinks = $('.govuk-summary-list__actions-list-item:last-of-type .govuk-visually-hidden', $rows);

      expect($rows.children().length).toBe(2);
      expect($identifiers.eq(0).text()).toContain('Eitem un');
      expect($changeLinks.eq(0).text()).toContain('Newid Eitem un');
      // TODO: add test for the change url
      expect($removeLinks.eq(0).text()).toContain('Dileu’r Eitem un o’r rhestr');
      // TODO: add test for the remove url

      expect($identifiers.eq(1).text()).toContain('Eitem dau');
      expect($changeLinks.eq(1).text()).toContain('Newid Eitem dau');
      expect($removeLinks.eq(1).text()).toContain('Dileu’r Eitem dau o’r rhestr');
    });

    it('has the item type in the legend text', () => {
      expect($legend.text()).toContain('Oes angen i chi ychwanegu eitem arall?');
    });

    it('has an action set on the form', () => {
      expect($form.attr('action')).toBe('#addItem');
    });
  });

  describe('Empty, null and missing itemLists', () => {
    function overrideItemList(itemList) {
      const params = { ...examples.default };
      params.itemList = itemList;
      return params;
    }

    it('should have the same output for all versions of no items being provided', () => {
      const outputs = [undefined, null, []].map((itemList) => render('add-to-a-list', overrideItemList(itemList)).html());

      expect(outputs[0]).toEqual(outputs[1]);
      expect(outputs[0]).toEqual(outputs[2]);
    });
  });
});
