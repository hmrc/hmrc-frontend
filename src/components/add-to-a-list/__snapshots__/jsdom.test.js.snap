// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add to a list should be rendered using a structure accessible for JAWS users 1`] = `

<div class="govuk-summary-list__row">
  <dt class="govuk-summary-list__key govuk-!-font-weight-regular hmrc-summary-list__key">
    Director One
  </dt>
  <dd class="govuk-summary-list__actions hmrc-summary-list__actions">
    <ul class="govuk-summary-list__actions-list">
      <li class="govuk-summary-list__actions-list-item">
        <a class="govuk-link"
           href="#change-item-1"
        >
          <span aria-hidden="true">
            Change
          </span>
          <span class="govuk-visually-hidden">
            Change Director One
          </span>
        </a>
      </li>
      <li class="govuk-summary-list__actions-list-item">
        <a class="govuk-link"
           href="#remove-item-1"
        >
          <span aria-hidden="true">
            Remove
          </span>
          <span class="govuk-visually-hidden">
            Remove Director One from the list
          </span>
        </a>
      </li>
    </ul>
  </dd>
</div>
<div class="govuk-summary-list__row">
  <dt class="govuk-summary-list__key govuk-!-font-weight-regular hmrc-summary-list__key">
    Director Two
  </dt>
  <dd class="govuk-summary-list__actions hmrc-summary-list__actions">
    <ul class="govuk-summary-list__actions-list">
      <li class="govuk-summary-list__actions-list-item">
        <a class="govuk-link"
           href="#change-item-2"
        >
          <span aria-hidden="true">
            Change
          </span>
          <span class="govuk-visually-hidden">
            Change Director Two
          </span>
        </a>
      </li>
      <li class="govuk-summary-list__actions-list-item">
        <a class="govuk-link"
           href="#remove-item-2"
        >
          <span aria-hidden="true">
            Remove
          </span>
          <span class="govuk-visually-hidden">
            Remove Director Two from the list
          </span>
        </a>
      </li>
    </ul>
  </dd>
</div>

`;
